### ISSUE TITLE
FSM: no such a transition JOINED -> DONOR

### ISSUE THREAD

**Issue by thoutenbos** [codership/galera#443]
We use the innobackupex script for 15 minute incremental backups. Normally the donating node should return to SYNCED state after rejoining. Occasionally it does not and ends up in JOINED state instead. In the next state change the transition is unknown.

Possible fix for this issue in the Percona fork:
https://github.com/percona/galera/commit/c51109c44cb39c94cfa9644334a810d3aeee66d2

```170120  1:50:07 [Note] WSREP: Member 0.0 (mariadb01) desyncs itself from group
170120  1:50:07 [Note] WSREP: Shifting SYNCED -> DONOR/DESYNCED (TO: 63011)
170120  1:50:07 [Note] WSREP: Provider paused at 6e9cfa8b-d9a6-11e6-bce8-1b2cba0fe012:63011 (66189)
170120  1:50:08 [Note] WSREP: resuming provider at 66189
170120  1:50:08 [Note] WSREP: Provider resumed.
170120  1:50:08 [Note] WSREP: Member 0.0 (mariadb01) resyncs itself to group
170120  1:50:08 [Note] WSREP: Shifting DONOR/DESYNCED -> JOINED (TO: 63011)
170120  1:50:08 [Note] WSREP: Member 0.0 (mariadb01) synced with group.
170120  1:50:08 [Note] WSREP: Shifting JOINED -> SYNCED (TO: 63011)
170120  1:50:08 [Note] WSREP: Synchronized with group, ready for connections


170120  2:05:11 [Note] WSREP: Member 0.0 (mariadb01) desyncs itself from group
170120  2:05:11 [Note] WSREP: Shifting SYNCED -> DONOR/DESYNCED (TO: 63110)
170120  2:05:11 [Note] WSREP: Provider paused at 6e9cfa8b-d9a6-11e6-bce8-1b2cba0fe012:63110 (66293)
170120  2:05:12 [Note] WSREP: resuming provider at 66293
170120  2:05:12 [Note] WSREP: Provider resumed.
170120  2:05:12 [Note] WSREP: Member 0.0 (mariadb01) resyncs itself to group
170120  2:05:12 [Note] WSREP: Shifting DONOR/DESYNCED -> JOINED (TO: 63110)


170120  2:20:08 [Note] WSREP: Member 0.0 (mariadb01) desyncs itself from group
170120  2:20:08 [Note] WSREP: Shifting JOINED -> DONOR/DESYNCED (TO: 63203)
170120  2:20:08 [ERROR] WSREP: FSM: no such a transition JOINED -> DONOR
170120  2:20:08 [ERROR] mysqld got signal 6 ;
This could be because you hit a bug. It is also possible that this binary
or one of the libraries it was linked against is corrupt, improperly built,
or misconfigured. This error can also be caused by malfunctioning hardware.

To report this bug, see https://mariadb.com/kb/en/reporting-bugs

We will try our best to scrape up some info that will hopefully help
diagnose the problem, but since we have already crashed,
something is definitely wrong and this may fail.

Server version: 10.0.28-MariaDB-wsrep
key_buffer_size=134217728
read_buffer_size=2097152
max_used_connections=11
max_threads=1002
thread_count=18
It is possible that mysqld could use up to
key_buffer_size + (read_buffer_size + sort_buffer_size)*max_threads = 6307645 K  bytes of memory
Hope that's ok; if not, decrease some variables in the equation.

Thread pointer: 0x0x7f9233050008
Attempting backtrace. You can use the following information to find out
where mysqld died. If you see no messages after this, something went
terribly wrong...
stack_bottom = 0x7f937b1b4d10 thread_stack 0x48000
/usr/sbin/mysqld(my_print_stacktrace+0x2e)[0xc0524e]
/usr/sbin/mysqld(handle_fatal_signal+0x3b6)[0x75d496]
/lib64/libpthread.so.0(+0xf370)[0x7f938f78e370]
/lib64/libc.so.6(gsignal+0x37)[0x7f938df401d7]
/lib64/libc.so.6(abort+0x148)[0x7f938df418c8]
/usr/lib64/galera/libgalera_smm.so(_ZN6galera3FSMINS_10Replicator5StateENS_13ReplicatorSMM10TransitionENS_10EmptyGuardENS_11EmptyActionEE8shift_toES2_+0x17c)[0x7f9387991a1c]
/usr/lib64/galera/libgalera_smm.so(_ZN6galera13ReplicatorSMM6desyncEv+0x70)[0x7f938798af10]
/usr/lib64/galera/libgalera_smm.so(galera_desync+0x19)[0x7f9387999ca9]
/usr/sbin/mysqld(_ZN16Global_read_lock34make_global_read_lock_block_commitEP3THD+0x35b)[0x82e19b]
/usr/sbin/mysqld(_Z20reload_acl_and_cacheP3THDyP10TABLE_LISTPi+0x33c)[0x6cd99c]
/usr/sbin/mysqld(_Z21mysql_execute_commandP3THD+0xa37)[0x5ddbc7]
/usr/sbin/mysqld[0x5e746d]
/usr/sbin/mysqld(_Z16dispatch_command19enum_server_commandP3THDPcj+0x1ce2)[0x5e9682]
/usr/sbin/mysqld(_Z10do_commandP3THD+0x2a2)[0x5ea432]
/usr/sbin/mysqld(_Z24do_handle_one_connectionP3THD+0x31b)[0x6b2b7b]
/usr/sbin/mysqld(handle_one_connection+0x39)[0x6b2c59]
/lib64/libpthread.so.0(+0x7dc5)[0x7f938f786dc5]
/lib64/libc.so.6(clone+0x6d)[0x7f938e00273d]

Trying to get some variables.
Some pointers may be invalid and cause the dump to abort.
Query (0x7f9230408020): is an invalid pointer
Connection ID (thread ID): 4418
Status: NOT_KILLED

Optimizer switch: index_merge=on,index_merge_union=on,index_merge_sort_union=on,index_merge_intersection=on,index_merge_sort_intersection=off,engine_condition_pushdown=off,index_condition_pushdown=on,derived_merge=on,derived_with_keys=on,firstmatch=on,loosescan=on,materialization=on,in_to_exists=on,semijoin=on,partial_match_rowid_merge=on,partial_match_table_scan=on,subquery_cache=on,mrr=off,mrr_cost_based=off,mrr_sort_keys=off,outer_join_with_cache=on,semijoin_with_cache=on,join_cache_incremental=on,join_cache_hashed=on,join_cache_bka=on,optimize_join_buffer_size=off,table_elimination=on,extended_keys=on,exists_to_in=on

The manual page at http://dev.mysql.com/doc/mysql/en/crashing.html contains
information that should help you find out what is causing the crash.

We think the query pointer is invalid, but we will try to print it anyway.
Query: FLUSH TABLES WITH READ LOCK```

**Comment by philip-galera** [comment]
@ayurchen , this issue seems to be a duplicate of GAL-480. Can you confirm that? If yes, then we have a fix already.

**Comment by thoutenbos** [comment]
@philip-galera @ayurchen any updates ?

**Comment by ayurchen** [comment]
Most likely it is a duplicate of GAL-480. In any case this particular FSM assertion has been removed so this symptom can't happen again. Fix released in 3.20.
